# Project 1-A Prime or Not a Prime

## Introduction

**Fermat's little theorem** is a basic fact that relates prime numbers and congruences: If $p$ is a prime number then $$a^p\equiv a (\mod p)$$ for any integer $0\leq a<p$.

For example, for $p=3$ we have
$$1^3=1\equiv 1 (\mod 3)$$
$$2^3=8\equiv 2 (\mod 3)$$
$$3^3=27\equiv 3 (\mod 3)$$
$$4^3=64\equiv 4 (\mod 3)$$
which shows that the formula $a^3\equiv a (\mod 3)$ holds for $a=1,2,3,4$.

The formula from the above theorem does not hold in general if $p$ is not prime. For example, for $p=4$ and $a=2$ we have $2^4=16$ which is not congruent to 2 modulo 4.

However, there are some numbers $p\geq 2$ such that:
* $p$ is not a prime
* the formula $a^p\equiv a(\mod p)$ holds for all $0\leq a<p$

We will call such numbers *false primes*.

In this project, our goal is to research the properties of false primes. The way to explore those is
1. to find out the first 20 false primes;
2. to decompose the first 20 false primes we found;
3. to conjecture the properties of false primes according to the decomposition.

## Part 1-First 20 False Primes

Write a Python script to find the first 20 false primes.

Step:
1. Write a Python function `is_prime` that checks whether the input number $n$ is prime or not
2. Write a Python function `is_prime_like` that checks whether Fermat's little theorem applies to the input number $n$ or not
3. Write a Python function `is_false_prime` that checks whether the input numbers $n$ is false prime or not
4. Create a list of first 20 false primes


We need to use the square root function for the first step. So, import the math module and store it as the variable `m` before step 1.

In [1]:
import math as m # import and simplify the math module name

### Step 1-Create a `is_prime` function:

A number is called prime when it is an integer greater than $1$, and the only positive integer factors of it are $1$ and the number itself. We can use the modulus operation `%` to check divisibility. If the reminder of `m` divided by `n` equals to $0$, then `n` will be a factor of `m`.

Given a number $n$, let every number from $2$ to $n-1$ divides $n$ then check if the reminder equals to $0$. If one number does, that means the one is a factor of $n$. Then we should `return false` since it is not prime. After running all of the numbers in the range from $2$ to $n-1$, no 'False' returns. It means that the given number $n$ is a prime, and then we should `return True`.

Optimization:
It takes a long time to run all numbers in the given range if the given number $n$ is huge. So we can replace $n$ with a least number that is necessary to check.

$$n = a\cdot b$$

Either $a\leq\sqrt{n}$ or $b\leq\sqrt{n}$

Hence, we substitube $\sqrt{n}+1$ for $n$ for the range upper bound. (Using '$\sqrt{n}+1$' instead of '$\sqrt{n}$' is because the last number is $\sqrt{n}-1$ if the range is from $2$ to $\sqrt{n}$.)

In [2]:
def is_prime(n):       
    if n < 2:                                # '0' and '1' cannot be used to divide any number to check if the given number is prime.
        return False                         # '0' and '1' is not prime, so return False.
    for d in range(2, (int(m.sqrt(n)) + 1)): # According to the above optimization, we need a for loop to let every number from '2' to 'sqrt(2)-1' divide the given number 'n'.
        if n % d == 0:                       # To check if the reminder of 'n' divieded by 'd' equals to '0'.
            return False                     # If the reminder equals to '0', that means 'd' is a factor of 'n'. Then 'n' is not prime, so return False.
    return True                              # Exist the loop, after checking all possible numbers, there's no False return. That means the given number 'n' is a prime. Return True.

To check if the `is_prime` function works for integer $2$.

In [3]:
is_prime(2)  # Apply the 'is_prime' function to integer '2'.

True

To check if the `is_prime` function works for integer $3$.

In [4]:
is_prime(3)  # Apply the 'is_prime' function to integer '3'.

True

To check if the `is_prime` function works for integer $4$, an integer that is not prime we can easily know.

In [5]:
is_prime(4)  # Apply the 'is_prime' function to integer '4'.

False

To check if the `is_prime` function works for integer $11$, an integer that is prime we can easily know.

In [6]:
is_prime(11)  # Apply the 'is_prime' function to integer '11'.

True

### Step 2-Create a `is_prime_like` function: 

We call a number $p$ *prime-like* if $p\geq 2$ and the formula $a^p \equiv a \pmod{p}$ holds for all $0\leq a <p$. Using this formula, we should check if the reminders from division of both $a^p$ and $a$ by $p$ are same, for all $a$ in range($2$, $n$). If not same, `return False`. Otherwise, `return True`.

In [7]:
def is_prime_like(n):
    for a in range(2,n):            # Set a loop for checking every number from '2' to 'n-1'
        if pow(a, n, n) != (a % n): # Check if the reminders from division of both 'a**p' and 'a' by 'p' are not the same
            return False
    return True

To check if the `is_prime_like` function works for integer $4$, which is a not prime-like example known from project introduction.

In [8]:
is_prime_like(4)  # Apply the 'is_prime_like' function to integer '4'.

False

To check if the `is_prime_like` function works for integer $561$, which is the smallest false prime number known from project introduction.

In [9]:
is_prime_like(561)  # Apply the 'is_prime_like' function to integer '561'.

True

### Step 3-Create a `is_false_prime` function

Combine the features of `is_prime_` function and `is_prime_like` function. It would turn out a function that can determine if the given number is false prime. When we input a number $n$, the `is_prime` function returns false and the `is_prime_like` function returns true, which means the $n$ is a false prime. Otherwise, it is not.

Note: _False primes_ are the numbers $p\geq 2$ such that:
1. _p_ is not a prime
2. the formula $a^p \equiv a \pmod{p}$ holds for all $0\leq a <p$

In [10]:
def is_false_prime(n):
    if not is_prime(n) and is_prime_like(n): # We don't want the input 'n' is a prime, and at the same time, 'n' is prime-like. 'not' will only apply on the condition after it.
        return True                          # If the input 'n' meets the above two conditions, return True.
    else:                      
        return False                         # If the input 'n' is prime, or it is not prime-like, return False.

### Step 4-Collect the first 20 false prime

Name an empty list `false_prime`. Since we don't know how mmany numbers we need to go through the loop to find the first 20 false primes, a `while` loop should be utilized here. Whenever we find a false prime, add it to the list of `false_prime`.

In [11]:
false_primes = []              # Set up an empty list named 'false prime'.

i = 2                          # Since false prime should be greater than '2', we start checking from '2'.

while len(false_primes) < 20:  # Set up a while loop. End runing while the length of the list of false primes is 20.
    if is_false_prime(i):      # Check if the input is false prime.
        false_primes.append(i) # If yes, add it to the list of false prime. 
    i += 1                     # Plus '1' each time the input is checked so that the while loop can proceed.

print(false_primes)            # Finish the collaection. Print out the list of false primes.

[561, 1105, 1729, 2465, 2821, 6601, 8911, 10585, 15841, 29341, 41041, 46657, 52633, 62745, 63973, 75361, 101101, 115921, 126217, 162401]


## Part 2-Primary decomposition of false prime

Exploring the properties of false primes is our goal of this project. In order to analyze them, we can decompose each false primes found above.

Step:
1. Write a Python function `my_primes` that returns the list of all primes smaller or equal to `n`.
2. Write a Python function `primary` that for each integer `n` greater than 1 returns a list $$[p_1,p_2,...,p_m]$$ where $p_1,...,p_m$ are primes such that $p_1\leq p_2\leq ...\leq p_m$ and $n=p_1\cdot p_2\cdot ...\cdot p_m$.
3. Show the primary decompositions of each false primes we found.

### Step 1-Create `my_primes` function

The function `is_prime` that can decide if the input is prime has been created. Now we use `is_prime` to find prime numbers that smaller than or equal to the given number `n`. Add the found prime numbers to a list named `prime_list`. Return this list.

In [12]:
def my_primes(n):
    prime_list = []              # Set up an empty list named 'prime_list'.
    for i in range(2,n+1):       # Check every number from '2' to 'n', so the upper bound should be 'n+1'.
        if is_prime(i):          # Apply the 'is_prime' function.
            prime_list.append(i) # Add the prime found to the list.
    return prime_list            # Return list after the loop ends.

To check if the `my_primes` function works for integer $11$.

In [13]:
my_primes(11)    # Apply the 'my_primes' function to integer '11'.

[2, 3, 5, 7, 11]

### Step 2-Create `primary` function

We still use the properties of prime numbers and congruences to create this function. Introduce the possible primes needed and let them be a range first. Next, set up a loop that check if every possible primes divide the input `n`. If yes, the primes will be a factor of `n`. Then add the prime factor to a list named `prime_decomposition`. Finally return the list.

In [14]:
def primary(n):
    primes = my_primes(n)                 # Introduce every primes smaller than or equal to 'n'.
    
    prime_decomposition = []              # Initialize an empty list.
    
    for d in primes:                      # Iterate numbers from 'primes'.
        while n % d == 0:                 # Set up a while loop to check if 'd' divides 'n'.
            prime_decomposition.append(d) # If yes, add 'd' to the list of 'prime_deconposition'.
            n /= d                        # Factor out 'd' from 'n' and get the new 'n'.
            
    return prime_decomposition            # Return list after the loop ends.

To check the `primary` function if works for a prime $11$.

In [15]:
primary(11) # Apply the 'primary' function to prime '11'.

[11]

To check the `primary` function if works for a non-prime number $10$.

In [16]:
primary(10) # Apply the 'primary' function to integer '10'.

[2, 5]

To test whether the `primary` function can decompose all prime factors without dropping any repeated.

In [17]:
primary(90) # Apply the 'primary' function to integer '90'.

[2, 3, 3, 5]

### Step 3-Primary decomposition of the first 20 false primes

Utilize a `for` loop and the `primary` function to print out the primary decomposition of the first 20 false primes we found above.

In [18]:
for x in false_primes:                                                     # Iterate each false prime from list of 'false_primes'
    print('The primary decomposition of {0} is {1}'.format(x, primary(x))) # Print out the decomposition results.

The primary decomposition of 561 is [3, 11, 17]
The primary decomposition of 1105 is [5, 13, 17]
The primary decomposition of 1729 is [7, 13, 19]
The primary decomposition of 2465 is [5, 17, 29]
The primary decomposition of 2821 is [7, 13, 31]
The primary decomposition of 6601 is [7, 23, 41]
The primary decomposition of 8911 is [7, 19, 67]
The primary decomposition of 10585 is [5, 29, 73]
The primary decomposition of 15841 is [7, 31, 73]
The primary decomposition of 29341 is [13, 37, 61]
The primary decomposition of 41041 is [7, 11, 13, 41]
The primary decomposition of 46657 is [13, 37, 97]
The primary decomposition of 52633 is [7, 73, 103]
The primary decomposition of 62745 is [3, 5, 47, 89]
The primary decomposition of 63973 is [7, 13, 19, 37]
The primary decomposition of 75361 is [11, 13, 17, 31]
The primary decomposition of 101101 is [7, 11, 13, 101]
The primary decomposition of 115921 is [13, 37, 241]
The primary decomposition of 126217 is [7, 13, 19, 73]
The primary decomposition

## Part 3-Conjecture

My conjecture about the properties of false primes:

According to the above observation, we can obviously see that all of the false primes we found are odd and composite. Because all primes are odd except $2$ and the false primes are composited by primes but $2$, I will conjecture all of the false primes are odd.

The difference between two adjacent false primes becomes larger as the false primes increasing. In addition, false primes are rare. It leads to some very huge numbers after the first 20 ones.

## Conclusion

In this project, in order to research the properties of one kind of numbers that are unfamiliar to the general public, we deeply utilize the features of prime numbers and congruences to analyze them. However, find out a number which is not a prime but was hold for the Fermat's little theorem is not an easy thing. In virtue of the power of computer and Python, we can calculate and test based on our demand in a short period of time. 

Above all, we need to write out the following functions: 1) `is_prime` function, which determine whether the input is prime; 2) `is_prime_like` function, which decide whether the input can pass the Fermat's little theorem. Moreoevr, combine the two functions above to create a new function `is_false_prime` that can determine whether the input is false prime. 

Secondly, decompose the false primes we found by using the `my_primes` and `primary` functions.

In the end, observe the results, and try to sperate them on each other, to see if we can come up with any patterns.

Results:
1. The first 20 false primes: \[561, 1105, 1729, 2465, 2821, 6601, 8911, 10585, 15841, 29341, 41041, 46657, 52633, 62745, 63973, 75361, 101101, 115921, 126217, 162401\]

2. The primary decomposition of 561 is \[3, 11, 17\]

   The primary decomposition of 1105 is \[5, 13, 17\]
   
   The primary decomposition of 1729 is \[7, 13, 19\]
   
   The primary decomposition of 2465 is \[5, 17, 29\]
   
   The primary decomposition of 2821 is \[7, 13, 31\]
   
   The primary decomposition of 6601 is \[7, 23, 41\]
   
   The primary decomposition of 8911 is \[7, 19, 67\]
   
   The primary decomposition of 10585 is \[5, 29, 73\]
   
   The primary decomposition of 15841 is \[7, 31, 73\]
   
   The primary decomposition of 29341 is \[13, 37, 61\]
   
   The primary decomposition of 41041 is \[7, 11, 13, 41\]
   
   The primary decomposition of 46657 is \[13, 37, 97\]
   
   The primary decomposition of 52633 is \[7, 73, 103\]
   
   The primary decomposition of 62745 is \[3, 5, 47, 89\]
   
   The primary decomposition of 63973 is \[7, 13, 19, 37\]
   
   The primary decomposition of 75361 is \[11, 13, 17, 31\]
   
   The primary decomposition of 101101 is \[7, 11, 13, 101\]
   
   The primary decomposition of 115921 is \[13, 37, 241\]
   
   The primary decomposition of 126217 is \[7, 13, 19, 73\]
   
   The primary decomposition of 162401 is \[17, 41, 233\]

3. My conjecture about the properties of false primes: all of the false primes are odd and composite. And they are all huge number.


## References

https://jllottes.github.io/Projects/prime_or_not/prime_or_not.html